### PR TITLE
[FLINK-25544][streaming][JUnit5 Migration] The runtime package of module flink-stream-java

### PR DESCRIPTION
**Pull request by Jiabao-Sun**
## What is the purpose of the change

[FLINK-25544][streaming][JUnit5 Migration] The runtime package of module flink-stream-java

## Brief change log

**1. Migrate Junit4 annotations to JUnit5**

JUnit 4 | JUnit 5
-- | --
`@Test` | `@Test`
`@Test(expected=XXX)` | `@Test + assertThatThrownBy(() -> { ...}).isInstanceOf(XXX.class)`
`@Test(timeout=XXX)` | `@Test + @Timeout(XXX)`
`@Before` | `@BeforeEach`
`@After` | `@AfterEach`
`@BeforeClass` | `@BeforeAll`
`@AfterClass` | `@AfterAll`
`@Ignore` | `@Disabled`
`@Category` | `@Tag`

**2. Migrate Junit4 Assertion & Assumption to AssertJ**
**3. Migrate Junit4 `@Rule` and `@ClassRule` to `@ExtendWith` and `@RegisterExtension`**
**4. Migrate Parameterized Test from `@RunWith(Parameterized.class)` to `@ExtendWith(ParameterizedTestExtension.class)` and `@TestTemplate`**
**4. Minimize Test classes and Methods visibility**

https://docs.google.com/document/d/1514Wa_aNB9bJUen4xm5uiuXOooOJTtXqS_Jqk9KJitU/edit?pli=1

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not documented)

### CODE REVIEW

**Comment by Jiabao-Sun** [comment]
Hi @ferenc-csaky, @GOODBOY008.
Could you help review this when you have time?
This is the last PR of flink-stream-java module.

**Comment by Jiabao-Sun** [comment]
Tests count verified.

master
Tests      run: 2353, Failures: 0, Errors: 0, Skipped: 132
IT Tests   run: 64, Failures: 0, Errors: 0, Skipped: 0

PR
Tests      run: 2353, Failures: 0, Errors: 0, Skipped: 132
IT   Tests run: 64, Failures: 0, Errors: 0, Skipped: 0

**Comment by Jiabao-Sun** [comment]
Thanks @JingGe for the suggestions.
Brief change log Updated.

**Comment by ferenc-csaky** [comment]
@Jiabao-Sun  FYI: I think I'll be finished with the review today or tomorrow.

**Review by JingGe — APPROVED** [review]
LGTM

**Comment by JingGe** [comment]
> Brief change log

@Jiabao-Sun Thanks for the excellent summary! I have added it as the reference example in the description of [FLINK-25325](https://issues.apache.org/jira/browse/FLINK-25325)

**Review by ferenc-csaky — CHANGES_REQUESTED** [review]
Thanks for pushing this forward! Added some comments/suggestions.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/StreamPartitionerTest.java [hunk 1]
```diff
@@ -21,33 +21,32 @@
...
-public abstract class StreamPartitionerTest extends TestLogger {
+abstract class StreamPartitionerTest {
 
     protected final StreamPartitioner<Tuple> streamPartitioner = createPartitioner();
     protected final StreamRecord<Tuple> streamRecord = new StreamRecord<>(null);
     protected final SerializationDelegate<StreamRecord<Tuple>> serializationDelegate =
             new SerializationDelegate<>(null);
 
-    abstract StreamPartitioner<Tuple> createPartitioner();
+    protected abstract StreamPartitioner<Tuple> createPartitioner();
```
[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/partitioner/StreamPartitionerTest.java:40]: Good point!

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TumblingProcessingTimeWindowsTest.java [hunk 2]
```diff
@@ -151,53 +132,44 @@ public void testTimeUnits() {
...
-            TumblingProcessingTimeWindows.of(Time.seconds(10), Time.seconds(-11));
-            fail("should fail");
-        } catch (IllegalArgumentException e) {
-            assertThat(e.toString(), containsString("abs(offset) < size"));
-        }
+    void testInvalidParameters() {
+
+        assertThatThrownBy(() -> TumblingProcessingTimeWindows.of(Time.seconds(-1)))
+                .isInstanceOf(IllegalArgumentException.class)
+                .hasMessageContaining("TumblingProcessingTimeWindows");
```
[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TumblingProcessingTimeWindowsTest.java:152]: Thanks.
I'll change the usage of deprecated `Time` to `Duration` as well.

[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TumblingProcessingTimeWindowsTest.java:152]: I found that in many test classes contain Time, I think it should be uniformly removed in Flink 2.0.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java [hunk 3]
```diff
@@ -895,39 +890,38 @@ public TaskMetricGroup getMetricGroup() {
                 (Gauge<Long>)
                         chainedOperatorMetricGroup.get(MetricNames.IO_CURRENT_OUTPUT_WATERMARK);
 
-        Assert.assertEquals(
-                "A metric was registered multiple times.",
-                5,
-                new HashSet<>(
+        assertThat(
```
[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java:893]: > the only reason for collecting the retrieved Gauge instances to a Set is to validate that none of them are null

Here, we are not checking for null; rather, we are examining whether there are duplicates among the five `Gauge`s to ensure that no metric has been registered multiple times.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/MergingWindowSetTest.java [hunk 4]
```diff
@@ -227,97 +220,92 @@ public void testLateMerging() throws Exception {
...
-                        containsInAnyOrder(new TimeWindow(10, 13))));
-        assertThat(mergeFunction.mergedStateWindows(), not(hasItem(mergeFunction.mergeTarget())));
-
-        assertThat(
-                windowSet.getStateWindow(new TimeWindow(0, 13)),
-                anyOf(
-                        is(new TimeWindow(0, 3)),
-                        is(new TimeWindow(5, 8)),
-                        is(new TimeWindow(10, 13))));
+        assertThat(windowSet.addWindow(new TimeWindow(0, 13), mergeFunction))
```
[the remark this answers is not in the document]
Comment (Jiabao-Sun) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/MergingWindowSetTest.java:285]: Good catch!

**Comment by Jiabao-Sun** [comment]
Thanks @ferenc-csaky for the hard review.
PTAL again when you have time.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (ferenc-csaky) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/OneInputStreamTaskTest.java:893]: Right, I misinterpreted this part, but even the equals message clearly tells what happens here. :)

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TumblingProcessingTimeWindowsTest.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (ferenc-csaky) [reply flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/operators/windowing/TumblingProcessingTimeWindowsTest.java]: Fair point, makes sense. 👍

**Review by ferenc-csaky — APPROVED** [review]
Thanks for the quick updates @Jiabao-Sun, LGTM!

**Pull request merged**